tensorboard/loader.py

```diff
@@ -0,0 +1,414 @@
+    self._clock = clock
+    self._is_started = False
+    self._is_closed = False
+    self._is_tapped = False
+    self._size = -1
+    self._last_stat = 0.0
+    self._buffered = 0
+    self._reader = RecordReader(self.path, start_offset)
+    self._records = collections.deque()  # type: collections.deque[Record]
+    self._read_exception = None  # type: Exception
+    self._close_exception = None  # type: Exception
+    self._lock = threading.Lock()
+    self._please_sir_i_want_some_more = threading.Condition(self._lock)
+    self._no_soup_for_you = threading.Condition(self._lock)
+    self._thread = threading.Thread(target=self._run,
+                                    name=_shorten_event_log_path(self.path))
+
+  def get_size(self):
+    """Returns byte length of file.
+
+    This is guaranteed to return a number greater than or equal to the
+    offset of the last record returned by get_next_record().
+
+    In the average case, this method will not block. However, if the
+    i/o thread has not yet computed this value, then this method will
+    block on a stat call.
+
+    This method can be called after the instance has been closed.
+
+    Returns:
+      The byte length of file, which might increase over time, but is
+      guaranteed to never decrease. It's also guaranteed that it will
+      be greater than or equal to the offset field of any Record.
+
+    Raises:
+      IOError: If the stat call failed.
+
+    :rtype: int
+    """
+    with self._lock:
+      size = self._size
+    if size == -1:
+      size = self._reader.get_size()
+      with self._lock:
+        if size > self._size:
+          self._size = size
+    return size
+
+  def get_next_record(self):
+    """Reads one record.
+
+    When this method is first called, it will spawn the thread and
+    block until a record is read. Once the thread starts, it will queue
+    up records which can be read without blocking. The exception is
+    when we reach the end of the file, in which case each repeated call
+    will be synchronous. There is no background polling. If new data is
+    appended to the file, new records won't be buffered until this
+    method is invoked again. The caller should take care to meter calls
+    to this method once it reaches the end of file, lest they impact
+    performance.
+
+    Returns:
+      A Record object, or None if there are no more records available
+      at the moment.
+
+    Raises:
+      IOError: If this instance has been closed.
+      Exception: To propagate any exceptions that may have been thrown
+          by the read operation in the other thread. If an exception is
+          thrown, then all subsequent calls to this method will rethrow
+          that same exception.
+
+    :rtype: Record
+    """
+    with self._lock:
+      if self._is_closed:
+        raise IOError('%s is closed' % self)
+      if not self._is_started:
+        self._thread.start()
+        self._is_started = True
```

Comment:
Per the [threading docs](https://docs.python.org/2/library/threading.html):

> Once the thread’s activity is started, the thread is considered ‘alive’. It stops being alive when its run() method terminates – either normally, or by raising an unhandled exception. The is_alive() method tests whether the thread is alive.

Since self._is_started initializes at False and changes only when the thread is started, I think it can be removed and you can check thread.is_alive instead.